android/engine/src/main/java/org/smartregister/fhircore/engine/rulesengine/RulesFactory.kt

```diff
@@ -774,6 +778,39 @@ constructor(
         }
       }
     }
+
+    fun getResourceByReference(
+      resourceReference: String?,
+    ): Resource? {
+      if (resourceReference.isNullOrEmpty()) {
+        return null
+      }
+
+      return runBlocking(dispatcherProvider.io()) {
+        try {
+          defaultRepository.loadResource(Reference().apply { reference = resourceReference })
+        } catch (e: ResourceNotFoundException) {
+          null
+        }
+      }
+    }
+
+    fun getResourceByIdAndType(
+      resourceId: String?,
+      resourceType: String?,
+    ): Resource? {
+      if (resourceId.isNullOrEmpty() || resourceType.isNullOrEmpty()) {
+        return null
+      }
+
+      return runBlocking(dispatcherProvider.io()) {
```

Comment:
This will block the main thread where the rules are executed. Use a new CoroutineScope instead.